File: backend/app/api/exam.py
import json
import logging
from fastapi import APIRouter, File, Form, HTTPException, Query, UploadFile
from starlette.concurrency import run_in_threadpool
from app.models.schemas import ExamStartRequest, ExamAnswerRequest, ExamAdvanceRequest
from app.services.exam_service import (
    create_session,
    get_examiner_intro,
    get_next_question,
    advance_session,
    ExamError,
)
from app.services.scoring_service import (
    generate_score_report,
    ScoringError,
    ScoringProviderError,
    ScoringSessionNotFoundError,
    ReportNotReadyError,
)
from app.services.data_loader import get_registry, DataError, InvalidExamError, validate_exam_id
from app.services import memory_store
from app.services.exam_audio_service import analyze_completed_exam, ExamAudioError


logger = logging.getLogger("api.exam")
router = APIRouter(prefix="/exam", tags=["exam"])
MAX_EXAM_AUDIO_BYTES = 100 * 1024 * 1024
MAX_EXAM_RECORDINGS = 20


@router.get("/memory")
def get_memory_summary():
    return memory_store.get_exam_memory_summary()


@router.get("/memory/{session_id}")
def get_memory_attempt(session_id: str):
    memory = memory_store.get_exam_memory(session_id)
    if not memory:
        raise HTTPException(status_code=404, detail="Saved exam memory not found.")
    return memory


@router.delete("/memory/{session_id}")
def delete_memory_attempt(session_id: str):
    if not memory_store.delete_exam_memory(session_id):
        raise HTTPException(status_code=404, detail="Saved exam memory not found.")
    return {"session_id": session_id, "deleted": True}


@router.get("/exams")
def list_exams(language: str = Query(None, description="Filter exams by training language")):
    try:
        registry = get_registry()
        exams = registry.get("exams", [])
        if language:
            exams = [e for e in exams if e.get("language", "") == language]
        return {"exams": exams}
    except DataError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        logger.error("Failed to list exams: %s", e)
        raise HTTPException(status_code=500, detail="Failed to load exam list. Please try again.")


@router.post("/start")
def start_exam(request: ExamStartRequest):
    try:
        validate_exam_id(request.exam_id)
        session_id = create_session(request.exam_id, "exam")
        intro = get_examiner_intro(session_id)
        return {
            "session_id": session_id,
            "examiner_message": intro,
            "current_part": "identity",
            "question_index": 0,
            "exam_id": request.exam_id,
        }
    except HTTPException:
        raise
    except InvalidExamError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except DataError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except ExamError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error("Failed to start exam: %s", e)
        raise HTTPException(status_code=500, detail="Failed to start the exam. Please try again.")


@router.post("/answer")
def submit_answer(request: ExamAnswerRequest):
    try:
        result = get_next_question(request.session_id, request.answer)
        return result
    except ExamError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except DataError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        logger.error("Failed to process answer: %s", e)
        raise HTTPException(status_code=500, detail="Failed to process your answer. Please try again.")


@router.post("/advance")
def advance_exam(request: ExamAdvanceRequest):
    try:
        return advance_session(request.session_id)
    except ExamError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except DataError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        logger.error("Failed to advance exam: %s", e)
        raise HTTPException(status_code=500, detail="Failed to advance the exam. Please try again.")


@router.post("/analysis/{session_id}")
async def analyze_exam_audio(
    session_id: str,
    files: list[UploadFile] = File(...),
    stages: str = Form(...),
    answer_indices: str = Form(...),
    language: str = Form("en"),
):
    """Run scoring-only audio analysis after the live test has finished."""
    if not files or len(files) > MAX_EXAM_RECORDINGS:
        raise HTTPException(status_code=400, detail="Invalid number of exam recordings.")
    try:
        parsed_stages = json.loads(stages)
        parsed_indices = json.loads(answer_indices)
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid recording stage metadata.")
    if (
        not isinstance(parsed_stages, list)
        or not isinstance(parsed_indices, list)
        or len(parsed_stages) != len(files)
        or len(parsed_indices) != len(files)
    ):
        raise HTTPException(status_code=400, detail="Recording stages do not match uploaded files.")

    recordings = []
    total_bytes = 0
    for upload, stage, answer_index in zip(files, parsed_stages, parsed_indices):
        remaining = MAX_EXAM_AUDIO_BYTES - total_bytes
        audio = await upload.read(remaining + 1)
        total_bytes += len(audio)
        if total_bytes > MAX_EXAM_AUDIO_BYTES:
            raise HTTPException(status_code=413, detail="Exam recordings exceed the 100 MB limit.")
        if audio:
            recordings.append({"stage": str(stage), "answer_index": int(answer_index), "audio": audio})
    try:
        result = await run_in_threadpool(analyze_completed_exam, session_id, recordings, language)
        return {
            "status": result["status"],
            "engine": result["engine"],
            "metrics": result["metrics"],
        }
    except ExamAudioError as exc:
        raise HTTPException(status_code=409, detail=str(exc))
    except RuntimeError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.get("/report/{session_id}")
def get_report(session_id: str):
    try:
        report = generate_score_report(session_id)
        return report
    except ScoringSessionNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ReportNotReadyError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except ScoringProviderError as e:
        raise HTTPException(status_code=502, detail=str(e))
    except ScoringError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except ExamError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except DataError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        logger.error("Failed to generate report: %s", e)
        raise HTTPException(status_code=500, detail="Failed to generate the score report. Please try again.")

File: frontend/src/api/exam.ts
import api from "./client";
import type { ExamMemorySummary } from "../types";

export interface ExamCueCard {
  topic: string;
  prompt_lines: string[];
  prep_seconds: number;
  speak_seconds: number;
}

export interface ExamStepResponse {
  next_question: string;
  is_finished: boolean;
  current_part: string;
  question_index: number;
  cue_card?: ExamCueCard;
}

export async function fetchExams() {
  const { data } = await api.get("/exam/exams");
  return data;
}

export async function startExam(examId: string = "ielts"): Promise<{
  session_id: string;
  examiner_message: string;
  current_part: string;
  question_index: number;
  exam_id: string;
}> {
  const { data } = await api.post("/exam/start", { exam_id: examId });
  return data;
}

export async function submitAnswer(sessionId: string, answer: string) {
  const { data } = await api.post("/exam/answer", {
    session_id: sessionId,
    answer,
  });
  return data as ExamStepResponse;
}

export async function advanceExam(sessionId: string) {
  const { data } = await api.post("/exam/advance", {
    session_id: sessionId,
  });
  return data as ExamStepResponse;
}

export interface ExamRecording {
  audio: Blob;
  stage: "part1" | "part2" | "part3";
  answerIndex: number;
}

export async function analyzeExamAudio(
  sessionId: string,
  recordings: ExamRecording[],
  language: string,
) {
  const body = new FormData();
  recordings.forEach((recording, index) => {
    body.append("files", recording.audio, `answer-${index + 1}.webm`);
  });
  body.append("stages", JSON.stringify(recordings.map((recording) => recording.stage)));
  body.append("answer_indices", JSON.stringify(recordings.map((recording) => recording.answerIndex)));
  body.append("language", language);
  const { data } = await api.post(`/exam/analysis/${sessionId}`, body, {
    headers: { "Content-Type": "multipart/form-data" },
    timeout: 0,
  });
  return data;
}

export async function getReport(sessionId: string) {
  const { data } = await api.get(`/exam/report/${sessionId}`);
  return data;
}

export async function getExamMemory(): Promise<ExamMemorySummary> {
  const { data } = await api.get("/exam/memory");
  return data;
}

export async function deleteExamMemory(sessionId: string): Promise<void> {
  await api.delete(`/exam/memory/${sessionId}`);
}
